### PR TITLE
New Adapter: Readpeak

### PR DESCRIPTION
**Pull request by readpeak-user**
Adds s2s adapter for Readpeak.

First time writing Go so please check carefully :)

Link to document PR https://github.com/prebid/prebid.github.io/pull/5242

### CODE REVIEW

**Comment by onkarvhanumante** [comment]
@readpeak-user 

tests are failing on PR with following error:

```
gofmt needs to be run, 6 files have issues.  Below is a list of files to review:
adapters/readpeak/params_test.go
adapters/readpeak/readpeak.go
adapters/readpeak/readpeak_test.go
exchange/adapter_builders.go
openrtb_ext/bidders.go
openrtb_ext/imp_readpeak.go
Error: Process completed with exit code 1.
```

run gofmt on above files to fix errors reported by test suite

File: static/bidder-info/readpeak.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+endpoint: https://dsp.readpeak.com/header/prebid
```
Comment (onkarvhanumante) [inline static/bidder-info/readpeak.yaml:1]: endpoint is reachable

```
curl -i --location --request POST 'https://dsp.readpeak.com/header/prebid'
HTTP/2 204
date: Fri, 05 Apr 2024 10:11:33 GMT
```

File: adapters/readpeak/readpeak.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+	endpoint string
+}
+
+// Builder builds a new instance of the Readpeak adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	bidder := &adapter{
+	  endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
```
Comment (onkarvhanumante) [inline adapters/readpeak/readpeak.go:27]: @readpeak-user  just a sugestion above code can be refactored as below.

```
func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
	return &adapter{
	  endpoint: config.Endpoint,
	}, nil
}

```

File: adapters/readpeak/readpeak.go [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+	for i := 0; i < len(requestCopy.Imp); i++ {
+		if requestCopy.Imp[i].Native == nil && requestCopy.Imp[i].Banner == nil {
+			continue
+		}
+		var impExt adapters.ExtImpBidder
+		err := json.Unmarshal(requestCopy.Imp[i].Ext, &impExt)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
```
Comment (onkarvhanumante) [inline adapters/readpeak/readpeak.go:44]: could be refacroed as below,

```
		var impExt adapters.ExtImpBidder
		if err := json.Unmarshal(requestCopy.Imp[i].Ext, &impExt); err != nil {
			errors = append(errors, err)
			continue
		}

```

File: adapters/readpeak/readpeak.go [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+		err2 := json.Unmarshal(impExt.Bidder, &rpExt)
+		if err2 != nil {
+			errors = append(errors, err2)
+			continue
+		}
```
Comment (onkarvhanumante) [inline adapters/readpeak/readpeak.go:49]: avoid use of `err2`. could be refactored as below.

```
		if err := json.Unmarshal(impExt.Bidder, &rpExt); err2 != nil {
			errors = append(errors, err)
			continue
		}

```

File: adapters/readpeak/readpeak.go [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+		return nil, []error{err}
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d.", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
```
Comment (onkarvhanumante) [inline adapters/readpeak/readpeak.go:123]: could make use of adapter response util here

https://github.com/prebid/prebid-server/blob/ec729e6e5d50a1306d76332fda52108b9d313c1d/adapters/response.go#L10-L28

File: adapters/readpeak/readpeak.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
```
Comment (onkarvhanumante) [inline adapters/readpeak/readpeak.go:131]: by default line 130 will initialize `bidResponse.Currency` as `USD`

reassigning `bidResponse.Currency` on line 131  without checking  `response.Cur` length may overwrite `bidResponse.Currency` with blank value

code block can be refactored as below

```
	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
        if len(response.Cur) != 0 {
	    bidResponse.Currency = response.Cur
        }
```

File: adapters/readpeak/readpeak.go [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+		bid.NURL = strings.Replace(bid.NURL, "${AUCTION_PRICE}", price, -1)
+		bid.AdM = strings.Replace(bid.AdM, "${AUCTION_PRICE}", price, -1)
+		bid.BURL = strings.Replace(bid.BURL, "${AUCTION_PRICE}", price, -1)
+	}
+}
+
+func getMediaType(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType, error) {
+	for _, imp := range imps {
+		if imp.ID == impID {
+			if imp.Banner != nil {
```
[the remark this answers is not in the document]
Comment (onkarvhanumante) [reply adapters/readpeak/readpeak.go:163]: @readpeak-user  PTAL at above comment and provide feedback on using `Mtype`

File: adapters/readpeak/readpeak.go [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+	if responseData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if responseData.StatusCode == http.StatusBadRequest {
+		err := &errortypes.BadInput{
+			Message: "Unexpected status code: 400. Bad request from publisher.",
+		}
+		return nil, []error{err}
+	}
```
Comment (onkarvhanumante) [inline adapters/readpeak/readpeak.go:116]: code path is not being covered

should add `supplemental` json test to cover above code path

File: adapters/readpeak/readpeak.go [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+				return openrtb_ext.BidTypeBanner, nil
+			} else if imp.Native != nil {
+				return openrtb_ext.BidTypeNative, nil
+			}
+		}
+	}
+
+	return "", &errortypes.BadInput{
+		Message: fmt.Sprintf("Failed to find impression type \"%s\"", impID),
+	}
```
Comment (onkarvhanumante) [inline adapters/readpeak/readpeak.go:173]: code path is not being covered

should add supplemental json test to cover above code path

File: static/bidder-info/readpeak.yaml [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+geoscope:
+  - EEA
+maintainer:
+  email: devteam@readpeak.com
```
Comment (onkarvhanumante) [inline static/bidder-info/readpeak.yaml:5]: @readpeak-user,  Email is being sent from prebid team to verify `devteam@readpeak.com` address. Requesting to respond back on email thread.

File: adapters/readpeak/readpeak_test.go [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+  "testing"
+
+  "github.com/prebid/prebid-server/v2/adapters/adapterstest"
+  "github.com/prebid/prebid-server/v2/config"
+  "github.com/prebid/prebid-server/v2/openrtb_ext"
+)
+
+func TestJsonSamples(t *testing.T) {
+  bidder, buildErr := Builder(openrtb_ext.BidderReadpeak, config.Adapter{
+    Endpoint: "http://readpeaktest.com/bidder"},
```
Comment (onkarvhanumante) [inline adapters/readpeak/readpeak_test.go:13]: should use same endpoint mentioned in bidder-info yaml

File: adapters/readpeak/readpeak.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (readpeak-user) [reply adapters/readpeak/readpeak.go]: Fixed to use mtype from bid response

File: adapters/readpeak/readpeak.go (hunk 2 above)
Comment (readpeak-user) [reply adapters/readpeak/readpeak.go]: Refactored. The original was copied from the docs here https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html.

File: adapters/readpeak/readpeak.go (hunk 4 above)
Comment (readpeak-user) [reply adapters/readpeak/readpeak.go]: Refactored

File: adapters/readpeak/readpeak.go (hunk 5 above)
Comment (readpeak-user) [reply adapters/readpeak/readpeak.go]: Refactored

File: adapters/readpeak/readpeak.go (hunk 6 above)
Comment (readpeak-user) [reply adapters/readpeak/readpeak.go]: Fixed

File: adapters/readpeak/readpeak.go (hunk 8 above)
Comment (readpeak-user) [reply adapters/readpeak/readpeak.go]: Added test

File: adapters/readpeak/readpeak.go (hunk 9 above)
Comment (readpeak-user) [reply adapters/readpeak/readpeak.go]: Added test

File: static/bidder-info/readpeak.yaml (hunk 10 above)
Comment (readpeak-user) [reply static/bidder-info/readpeak.yaml:5]: Replied

File: adapters/readpeak/readpeak_test.go (hunk 11 above)
Comment (readpeak-user) [reply adapters/readpeak/readpeak_test.go]: Updated the endpoint, although docs here https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html say you can use whatever URL.

**Comment by gargcreation1992** [comment]
@readpeak-user Pr checks are failing as you would need to run gofmt in your code files.

File: static/bidder-info/readpeak.yaml (hunk 10 above)
Comment (onkarvhanumante) [reply static/bidder-info/readpeak.yaml:5]: received reply from `devteam@readpeak.com`

File: adapters/readpeak/readpeak.go [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var errors []error
+
+	requestCopy := *request
+	var rpExt openrtb_ext.ImpExtReadpeak
+	var imps []openrtb2.Imp
+	for i := 0; i < len(requestCopy.Imp); i++ {
+		if requestCopy.Imp[i].Native == nil && requestCopy.Imp[i].Banner == nil {
+			continue
+		}
```
Comment (onkarvhanumante) [inline adapters/readpeak/readpeak.go:37]: `static/bidder-info/readpeak.yaml` specifies either `banner` or `Native` is required. 
prebid server core will ensure either `banner` or `native` is present else will return error before adapter code is being invoked

File: adapters/readpeak/readpeak.go [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+		Headers: headers,
+	}
+
+	return []*adapters.RequestData{requestData}, errors
+}
+
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if responseData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
```
Comment (onkarvhanumante) [inline adapters/readpeak/readpeak.go:106]: https://github.com/prebid/prebid-server/blob/ec729e6e5d50a1306d76332fda52108b9d313c1d/adapters/response.go#L26-L28

could make use of response util here

**Comment by onkarvhanumante** [comment]
@readpeak-user  test suites are failing with following error

```
--- FAIL: TestJsonSamples (0.00s)
    test_json.go:128: 
        	Error Trace:	/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:256
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:412
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:128
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:99
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:76
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:445
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:535
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:70
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/readpeak_test.go:20
        	Error:      	Received unexpected error:
        	            	expected.ImpIDs must contain at least one imp ID
        	Test:       	TestJsonSamples
        	Messages:   	readpeaktest/exemplary/banner.json Expected RequestData was not returned by adapters' MakeRequests() implementation: httpRequest[0]
    test_json.go:128: 
        	Error Trace:	/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:256
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:412
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:128
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:99
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:76
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:445
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:535
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:70
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/readpeak_test.go:20
        	Error:      	Received unexpected error:
        	            	expected.ImpIDs must contain at least one imp ID
        	Test:       	TestJsonSamples
        	Messages:   	readpeaktest/exemplary/banner_app.json Expected RequestData was not returned by adapters' MakeRequests() implementation: httpRequest[0]
    test_json.go:128: 
        	Error Trace:	/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:[256](https://github.com/prebid/prebid-server/actions/runs/8700881813/job/23871502985?pr=3610#step:4:257)
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:412
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:128
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:99
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:76
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:445
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:535
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:70
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/readpeak_test.go:20
        	Error:      	Received unexpected error:
        	            	expected.ImpIDs must contain at least one imp ID
        	Test:       	TestJsonSamples
        	Messages:   	readpeaktest/exemplary/banner_multiple_imps.json Expected RequestData was not returned by adapters' MakeRequests() implementation: httpRequest[0]
    test_json.go:128: 
        	Error Trace:	/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:256
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:412
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:128
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:99
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:76
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:445
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:535
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:70
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/readpeak_test.go:20
        	Error:      	Received unexpected error:
        	            	expected.ImpIDs must contain at least one imp ID
        	Test:       	TestJsonSamples
        	Messages:   	readpeaktest/exemplary/native.json Expected RequestData was not returned by adapters' MakeRequests() implementation: httpRequest[0]
    test_json.go:128: 
        	Error Trace:	/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:256
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:412
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:128
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:99
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:76
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:445
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:535
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:70
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/readpeak_test.go:20
        	Error:      	Received unexpected error:
        	            	expected.ImpIDs must contain at least one imp ID
        	Test:       	TestJsonSamples
        	Messages:   	readpeaktest/supplemental/banner_without_mtype.json Expected RequestData was not returned by adapters' MakeRequests() implementation: httpRequest[0]
    test_json.go:128: 
        	Error Trace:	/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:256
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:412
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:128
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:99
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:76
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:445
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:535
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:70
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/readpeak_test.go:20
        	Error:      	Received unexpected error:
        	            	expected.ImpIDs must contain at least one imp ID
        	Test:       	TestJsonSamples
        	Messages:   	readpeaktest/supplemental/bid_response_204.json Expected RequestData was not returned by adapters' MakeRequests() implementation: httpRequest[0]
    test_json.go:128: 
        	Error Trace:	/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:256
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:412
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:128
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:99
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:76
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:445
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:535
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:70
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/readpeak_test.go:20
        	Error:      	Received unexpected error:
        	            	expected.ImpIDs must contain at least one imp ID
        	Test:       	TestJsonSamples
        	Messages:   	readpeaktest/supplemental/bid_response_400.json Expected RequestData was not returned by adapters' MakeRequests() implementation: httpRequest[0]
    test_json.go:128: 
        	Error Trace:	/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:256
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:412
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:128
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:99
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:76
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:445
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/path.go:535
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/test_json.go:70
        	            				/home/runner/work/prebid-server/prebid-server/adapters/readpeak/readpeak_test.go:20
        	Error:      	Received unexpected error:
        	            	expected.ImpIDs must contain at least one imp ID
        	Test:       	TestJsonSamples
        	Messages:   	readpeaktest/supplemental/bid_response_500.json Expected RequestData was not returned by adapters' MakeRequests() implementation: httpRequest[0]
FAIL
	github.com/prebid/prebid-server/v2/adapters/readpeak	coverage: 92.1% of statements
FAIL	github.com/prebid/prebid-server/v2/adapters/readpeak	0.026s
```

File: adapters/readpeak/readpeak.go (hunk 12 above)
Comment (readpeak-user) [reply adapters/readpeak/readpeak.go]: Removed the check

File: adapters/readpeak/readpeak.go (hunk 13 above)
Comment (readpeak-user) [reply adapters/readpeak/readpeak.go]: Used util function

**Comment by readpeak-user** [comment]
When I run the tests locally with `./validate.sh --nofmt --cov --race 10`, they pass:

Screenshot 2024-04-18 at 12 07 58

**Comment by onkarvhanumante** [comment]
> @onkarvhanumante I have not been able to figure out why those tests fail. Any ideas? How can I run these tests locally?

@readpeak-user  requesting to merge `Prebid:master` changes into `readpeak:readpeak-adapter` and follow the pattern mentioned below

https://github.com/prebid/prebid-server/blob/763551963dd3208dde6f9bf35b31ef76cc83814f/adapters/pubmatic/pubmatic.go#L175-L182

**Comment by readpeak-user** [comment]
Thanks @onkarvhanumante. Should be fixed now.

**Review by gargcreation1992 — APPROVED** [review]

**Review by onkarvhanumante — APPROVED** [review]

**Pull request merged**